cpp/include/cuvs/neighbors/dynamic_batching.hpp

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuvs/neighbors/common.hpp>
+
+#include <cuvs/neighbors/cagra.hpp>
+#include <cuvs/neighbors/ivf_flat.hpp>
+#include <cuvs/neighbors/ivf_pq.hpp>
+
+namespace cuvs::neighbors::dynamic_batching {
+
+namespace detail {
+template <typename T, typename IdxT>
+class batch_runner;
+}
+
+template <typename Upstream>
+struct index_params : cuvs::neighbors::index_params {
+  /** The index, to which the dynamic batches are dispatched. */
+  Upstream& upstream;
+  /** Search paramerets for all requests within a batch. */
+  typename Upstream::search_params_type& upstream_params;
+  /** Filtering function, if any, must be the same for all requests in a batch. */
+  cuvs::neighbors::filtering::base_filter* sample_filter = nullptr;
+  /** The number of neighbors to search is fixed at construction time. */
+  int64_t k;
+  /** Input data (queries) dimensionality. */
+  int64_t dim;
```

Comment:
Could we remove `dim` and query the original index for `dim` while constructing the batched index?

Reply:
Hmm, probably yes. I'd need to check if all our indices have a matching getter method with exactly the same signature.